src/lib/sw_engine/tvgSwRenderer.cpp

```diff
@@ -76,7 +76,9 @@ struct SwShapeTask : SwTask
 
     void run(unsigned tid) override
     {
-        if (opacity == 0) return;  //Invisible
+        auto compMethod = CompositeMethod::None;
+        auto usedAsClip = sdata->composite(nullptr, &compMethod) && (compMethod == CompositeMethod::ClipPath);
```

Comment:
According to the code, it seems important to find out whether it is used as a clip in the SwRenderer.
I think this could be handled internally, but I can't come up with a good way.
um... I think we can check them because we update the clips first in the "prepare" step.
maybe like this?
```diif
--- a/src/lib/sw_engine/tvgSwRenderer.cpp
+++ b/src/lib/sw_engine/tvgSwRenderer.cpp
@@ -73,11 +73,12 @@ struct SwShapeTask : SwTask
     SwShape shape;
     const Shape* sdata = nullptr;
     bool cmpStroking = false;
+    bool usedAsClip = false;
 
     void run(unsigned tid) override
     {
         auto compMethod = CompositeMethod::None;
-        auto usedAsClip = sdata->composite(nullptr, &compMethod) && (compMethod == CompositeMethod::ClipPath);
+        //auto usedAsClip = sdata->composite(nullptr, &compMethod) && (compMethod == CompositeMethod::ClipPath);
         if (opacity == 0 && !usedAsClip) return;  //Invisible
 
         uint8_t strokeAlpha = 0;
@@ -593,6 +594,7 @@ void* SwRenderer::prepareCommon(SwTask* task, const RenderTransform* transform,
     if (clips.count > 0) {
         //Guarantee composition targets get ready.
         for (auto clip = clips.data; clip < (clips.data + clips.count); ++clip) {
+            static_cast<SwShapeTask*>(*clip)->usedAsClip = true;
             static_cast<SwShapeTask*>(*clip)->done();

```

If the composite() getter API could be used elsewhere, this might be better.